src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.PendingBranchesCollection.cs

```diff
@@ -0,0 +1,154 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class AbstractFlowPass<TLocalState, TLocalFunctionState>
+    {
+        internal sealed class PendingBranchesCollection
+        {
+            private ArrayBuilder<PendingBranch> _unlabeledBranches;
+            private PooledDictionary<LabelSymbol, ArrayBuilder<PendingBranch>>? _labeledBranches;
+
+            internal PendingBranchesCollection()
+            {
+                _unlabeledBranches = ArrayBuilder<PendingBranch>.GetInstance();
+            }
+
+            internal void Free()
+            {
+                _unlabeledBranches.Free();
+                _unlabeledBranches = null!;
+                FreeLabeledBranches();
+            }
+
+            internal void Clear()
+            {
+                _unlabeledBranches.Clear();
+                FreeLabeledBranches();
+            }
+
+            private void FreeLabeledBranches()
+            {
+                if (_labeledBranches is { })
+                {
+                    foreach (var branches in _labeledBranches.Values)
+                    {
+                        branches.Free();
+                    }
+                    _labeledBranches.Free();
+                    _labeledBranches = null;
+                }
+            }
+
+            /// <summary>
+            /// Returns the unordered collection of branches.
+            /// </summary>
+            internal ImmutableArray<PendingBranch> ToImmutable()
```

Comment:
>ToImmutable [](http://example.com/codeflow?start=51&length=11)

Where is this helper used now?

Reply:
`ToImmutable()` is called from `AbstractFlowPass.RemoveReturns()`.